File: services/api-gateway/app/schemas/inference.py
from typing import Any

from pydantic import BaseModel, Field


class RuntimePredictionSummary(BaseModel):
    label: str | None
    confidence: float | None = None
    model_name: str | None = None
    mode: str | None = None
    prediction_source: str | None = None
    is_fallback: bool


class RuntimeInferenceHistoryItem(BaseModel):
    id: str
    text: str
    sentiment: RuntimePredictionSummary
    aspect: RuntimePredictionSummary
    created_at: str


class RuntimeInferenceHistoryData(BaseModel):
    items: list[RuntimeInferenceHistoryItem] = Field(default_factory=list)
    total: int
    page: int = 1
    limit: int = 20
    total_pages: int = 1


class RuntimeInferenceResult(BaseModel):
    id: str
    text: str
    sentiment: dict[str, Any]
    aspect: dict[str, Any]
    saved: bool
    created_at: str

File: frontend/types/inference.ts
export interface RuntimeInferenceRequest {
  text: string;
}

export interface RuntimePredictionBase {
  label: string | null;
  confidence: number | null;
  model_name: string | null;
  mode: string | null;
  prediction_source: string | null;
  model_available: boolean;
  is_fallback: boolean;
}

export interface SentimentInferenceResult extends RuntimePredictionBase {
  probabilities: Record<string, number>;
}

export interface AspectInferenceResult extends RuntimePredictionBase {
  scores: Record<string, number>;
}

export interface RuntimeInferenceResult {
  id: string;
  text: string;
  sentiment: SentimentInferenceResult;
  aspect: AspectInferenceResult;
  saved: boolean;
  created_at: string;
  warnings?: string[];
  persistence_error?: {
    code: string;
    details?: string | Record<string, unknown>;
  };
}

export interface RuntimeInferenceHistoryItem {
  id: string;
  text: string;
  sentiment: SentimentInferenceResult;
  aspect: AspectInferenceResult;
  request_source?: string;
  created_at: string;
}

export interface RuntimeInferenceHistoryResponse {
  items: RuntimeInferenceHistoryItem[];
  total: number;
  page?: number;
  limit?: number;
  total_pages?: number;
}

export interface RuntimeInferenceHistoryQuery {
  limit?: number;
  page?: number;
}
